traffic_ops/traffic_ops_golang/deliveryservice/servers/servers.go

```diff
@@ -263,7 +243,428 @@ func (dss *TODeliveryServiceServer) Delete(db *sqlx.DB, user auth.CurrentUser) (
 	rollbackTransaction = false
 	return nil, tc.NoError
 }
-func selectQuery() string {
+
+func selectQuery(orderBy string, limit string, offset string) (string, error) {
+
+	selectStmt := `SELECT
+	s.deliveryService,
+	s.server,
+	s.last_updated
+	FROM deliveryservice_server s`
+
+	allowedOrderByCols := map[string]string{
+		"":                "",
+		"deliveryservice": "s.deliveryService",
+		"server":          "s.server",
+		"lastupdated":     "s.last_updated",
+		"deliveryService": "s.deliveryService",
+		"lastUpdated":     "s.last_updated",
+		"last_updated":    "s.last_updated",
+	}
+	orderBy, ok := allowedOrderByCols[orderBy]
+	if !ok {
+		return "", errors.New("orderBy '" + orderBy + "' not permitted")
+	}
+
+	if orderBy != "" {
+		selectStmt += ` ORDER BY ` + orderBy
+	}
+
+	selectStmt += ` LIMIT ` + limit + ` OFFSET ` + offset + ` ROWS`
+	return selectStmt, nil
+}
+
+func deleteQuery() string {
+	query := `DELETE FROM deliveryservice_server
+	WHERE deliveryservice=:deliveryservice and server=:server`
+	return query
+}
+
+type DSServers struct {
+	DsId    *int  `json:"dsId" db:"deliveryservice"`
+	Servers []int `json:"servers"`
+	Replace *bool `json:"replace"`
+}
+
+type TODSServers DSServers
+
+
+func createServersForDsIdRef() *TODSServers {
+	var dsserversRef = TODSServers(DSServers{})
+	return &dsserversRef
+}
+
+func GetReplaceHandler(db *sqlx.DB) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		defer r.Body.Close()
+		handleErrs := tc.GetHandleErrorsFunc(w, r)
+		ctx := r.Context()
+		user, err := auth.GetCurrentUser(ctx)
+		if err != nil {
+			log.Errorf("unable to retrieve current user from context: %s", err)
+			handleErrs(http.StatusInternalServerError, err)
+			return
+		}
+
+		// get list of server Ids to insert
+		payload :=  createServersForDsIdRef() 
+		servers := payload.Servers
+		dsId := payload.DsId
+
+		if err := json.NewDecoder(r.Body).Decode(payload); err != nil {
+			log.Errorf("Error trying to decode the request body: %s", err)
+			handleErrs(http.StatusInternalServerError, err)
+			return
+		}
+
+		// if the object has tenancy enabled, check that user is able to access the tenant
+		// check user tenancy access to this resource.
+		row := db.QueryRow("SELECT xml_id FROM deliveryservice WHERE id = $1", *dsId)
```

Comment:
This is panicing. It looks like it's because You're assigning `dsId` above before decoding the JSON, so it's always `nil`.

But, even if the JSON has been decoded, you still need to check for nil, because the user POST could be missing that field. In which case, we should return a `http.StatusBadRequest` to the user, with a message explaining that `dsId` is required. 

The `Replace` field also needs checked that it's not `nil` before dereferencing.